pkg/scheduler/framework/plugins/volumezone/volume_zone.go

```diff
@@ -261,22 +268,206 @@ func getErrorAsStatus(err error) *framework.Status {
 // EventsToRegister returns the possible events that may make a Pod
 // failed by this plugin schedulable.
 func (pl *VolumeZone) EventsToRegister() []framework.ClusterEventWithHint {
+
 	return []framework.ClusterEventWithHint{
 		// New storageClass with bind mode `VolumeBindingWaitForFirstConsumer` will make a pod schedulable.
 		// Due to immutable field `storageClass.volumeBindingMode`, storageClass update events are ignored.
-		{Event: framework.ClusterEvent{Resource: framework.StorageClass, ActionType: framework.Add}},
+		{Event: framework.ClusterEvent{Resource: framework.StorageClass, ActionType: framework.Add}, QueueingHintFn: pl.isSchedulableAfterStorageClassAdded},
 		// A new node or updating a node's volume zone labels may make a pod schedulable.
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.UpdateNodeLabel}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.UpdateNodeLabel}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
 		// A new pvc may make a pod schedulable.
 		// Due to fields are immutable except `spec.resources`, pvc update events are ignored.
-		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add}},
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add}, QueueingHintFn: pl.isSchedulableAfterPersistentVolumeClaimAdded},
 		// A new pv or updating a pv's volume zone labels may make a pod schedulable.
-		{Event: framework.ClusterEvent{Resource: framework.PersistentVolume, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolume, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterPersistentVolumeChange},
+	}
+}
+
+// isSchedulableAfterStorageClassAdded is invoked whenever a StorageClass added. It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeZone) isSchedulableAfterStorageClassAdded(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	_, addedStorageClass, err := util.As[*storage.StorageClass](nil, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterStorageClassAdded", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	for i := range pod.Spec.Volumes {
```

Comment:
These codes(L295-318) appeared in this file so many times... 
We should try to uniform them in a single function.
Maybe something like:
```
func (pl *VolumeZone) getPVsTopologies(ctx context.Context, pod *v1.Pod, matchPV, matchSC string) (pvTopologies []pvTopology, status *framework.Status, pvMatched bool, scMatched bool) {

}

```
WDYT?

Reply:
Thank you for your comment!
I understand your opinion well.

Indeed, I am writing similar code in each function(`isSchedulableAfterStorageClassAdded`, `isSchedulableAfterPersistentVolumeClaimAdded`, `getUpdatedPVsTopologies`), but in fact, what they are doing is slightly different. For example, `isSchedulableAfterStorageClassAdded` checks `StorageClass`'s name, but `isSchedulableAfterPersistentVolumeClaimAdded` checks `PVC`'s name.

Therefore, I thought it difficult to prepare another function separately as you suggested.

Reply:
> Therefore, I thought it difficult to prepare another function separately as you suggested.

Yes, but I think it's still needed for a clean code.
One easy way may be to separate these functions to server parts: 

1. get all pvc name from the pod 
2. do some operations on these pvc (like check the sc'name or check the pvc's name)
3. get pvtopology from pv bounded to these pvc

Then, we can separate two small functions which can be used in other 4 functions(getPVbyPod,isSchedulableAfterStorageClassAdded ...)
```go
// first use this function to get all pvc's names
func getPvScFromPod(pod *v1.Pod) (pvcNames []string)
```
```go
// then do something for check using these functions 
pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(pvcName)
storagehelpers.GetPersistentVolumeClaimClass(pvc)
```
```go
// after getting all the pv using the name slice returned by the first function, use this to get topology
func getPvTopologyFromPV(PV *v1.PersistentVolume) []pvTopology 
```

For example, `getUpdatedPVsTopologies` is now only checking the pv's name, and it gets the topologies from the pv which passed to the function, which is a little unintuitive :)

But I think we can leave it as is for now and wait for other reviewers' opinions.

Reply:
Thanks for your clarification!
Yeah I will wait for another reviewer's comment

Reply:
Hi @AxeZhan,

I refactored some functions, though code change might be different from your intention.
Could you review it again ?
If it looks good, I want to ask another reviewer